docs/specs/code/SIGNING.md

```diff
@@ -0,0 +1,183 @@
+
+### Proposed format
+
+##### Definitions
+
+**Define**: New CBOR tag `MessageTag = 0x0b` to be used for our new message signing standard. This tag is different than all other tags for meaningful signed data inside Cardano.
+
+**Define**: byte array `UserMessage` which is the message the user wishes to sign. It is RECOMMENDED to use unicode with utf8 encoding if the message is a string and RECOMMENDED to use network-byte-order for numbers.
+
+**Define**: function `BASE32` which converts bytes to base32 in the following way
+
+1) Consider byte array as an array of 5-bit values (most significant bit first)
+1) 0-pad last item if necessary
+1) Translate the array elements using the [bech32 table](https://github.com/bitcoin/bips/blob/master/bip-0173.mediawiki#bech32)
+
+**Define**: uint8 `KeyType` which exists so that programming languages can assign the correct type to these keys if needed. Although there is no standard specifying a number for each key type, in practice the following are used
+
+| Type            | ID |
+|-----------------|----|
+| XPub            | 0  |
+| ValidatorScript | 1  |
+| RedeemPublicKey | 2  |
+
+**Define**: uint32 `ProtocolMagic` which specifies the chain id. Notably, the `ProtocolMagic` for the Cardano Mainnet is `764824073`
+
+**Define**: function `SIGN` which cryptographically signs data. This function is purposely left vague as the implementation depends on what kind of key is used.
+
+**Define**: uint8 `FormatVersion` which specifies the version number for the proposed format.
+
+**Define**: byte* `Recipient` which is a identifier for who you are signing this message for. This is to avoid replay attacks as dapps for example can specify their contract address as the recipient. It is still not fool-proof as somebody can ask you to sign a message with the recipient set the same as the recipient field of the contract/person they intend to pass your message along to.
+
+##### FormatVersion 1
+
+```
+HRP             ::= 'cm'
+UserMessage     ::= byte*
+MessageTag      ::= 0x0b
+ProtocolMagic   ::= uint32
+Recipient       ::= byte*
+Signature       ::= SIGN(MessageTag || ProtocolMagic || UserMessage)
+                  | SIGN(MessageTag || ProtocolMagic || Recipient || UserMessage)
+
+KeyType         ::= uint8
+KeyData         ::= byte*
+PlainSignature ::= [ KeyType, Signature ]
+Witness         ::= [ KeyType, KeyData, Signature ]
+
+Signatures      ::= [ PlainSignature+ ]
+Witnesses       ::= [ Witness+ ]
+Elements        ::= 0 Signatures
+                  | 1 Witnesses
+
+FormatVersion   ::= uint8
+RawData         ::= BASE32([ FormatVersion , Elements ])
+                  | BASE32([ FormatVersion , Elements, Recipient ])
+
+DATA            ::= RawData CHECKSUM(HRP, RawData)
```

Comment:
How is CHECKSUM defined? Is this the same checksum defined in Bip-0173?

Reply:
yes it uses BCH